### PR TITLE
Restore hidden cursor after move completion (#2185)

### PR DESCRIPTION
**Pull request by ssoloff**
This PR addresses the cursor remaining hidden after move completion, as reported in #2185.  It also fixes the following related issues I observed during my investigation:

* The cancel button on the move panel has an empty label.
* The cancel button on the move panel was always enabled; it should be disabled when no move is in progress.

It appears that all of these issues are related to the changes made to `AbstractMovePanel` in 7c6df19.  Specifically:

* The new `cancelMoveEnabled` field had no effect on the enablement state of the cancel button in the UI.
* The `cancelMove()` method was changed to only execute its body if `cancelMoveEnabled` was `true`.  This was the root cause of the hidden cursor not being restored after completing a move because by the time `cancelMove()` was executed (via the call in `setActive()`), the `cancelMoveEnabled` flag had been reset to `false` due to the execution of `cancelMove()` being deferred via `SwingUtilities#invokeLater()`.

I fixed these issues by partially reverting some changes from 7c6df19:

* `cancelMoveEnabled` was removed and the original `cancelMoveAction` field was restored.
* Code to set the `cancelMoveEnabled` flag was replaced with the original code to enable/disable `cancelMoveAction`.
* `cancelMove()` was restored to always execute its body regardless of the enablement state of `cancelMoveAction`.  That is, when this method is called programatically, it will always be executed; execution will only be conditional when invoked via clicking the cancel button.

Please advise if any improvements can be made to address the spirit of the changes from 7c6df19.

NB: I think it's appropriate to squash the commits in this PR before merging.  They all essentially fix different symptoms of the same root cause.

### CODE REVIEW

File: src/main/java/games/strategy/ui/SwingComponents.java [hunk 1]
```diff
@@ -66,15 +66,15 @@
...
+   *   AbstractAction action = SwingComponents.newAbstractAction("Something", () -> {
    *    if (value) {
    *      doSomething();
    *    }
    *   });
    * </pre></code>
    */
-  public static AbstractAction newAbstractAction(final Runnable action) {
-    return new AbstractAction() {
+  public static AbstractAction newAbstractAction(final String name, final Runnable action) {
```
Comment (DanVanAtta) [inline src/main/java/games/strategy/ui/SwingComponents.java]: Is the `name` parameter value ignored or does it overwrite the buttons' name?  Would be nice to see that answered in the javadoc. Though this makes me wonder, `SwingComponents` was supposed to be *the* API Swing was not, instead we are seemingly just wrapping a crappy Swing API with one that allows for lambda's. `JButtonBuilder` solves these problems, my suggestion is to remove `newAbstractAction` completely from `SwingComponents`

File: src/main/java/games/strategy/triplea/ui/AbstractMovePanel.java [hunk 2]
```diff
@@ -244,7 +245,7 @@ protected final void display(final PlayerID id, final String actionLabel) {
       this.actionLabel.setText(id.getName() + actionLabel);
       add(leftBox(this.actionLabel));
       if (setCancelButton()) {
-        add(leftBox(new JButton(SwingComponents.newAbstractAction(this::cancelMove))));
+        add(leftBox(new JButton(cancelMoveAction)));
```
Comment (DanVanAtta) [inline src/main/java/games/strategy/triplea/ui/AbstractMovePanel.java]: Would it work to store a `JButton` instead of an `AbstractAction`. That avoids the redundancy of specifying the button title both as part of the button and again as part of the abstract action, and would avoid new'ing up a JButton here.

File: src/main/java/games/strategy/ui/SwingComponents.java (hunk 1 above)
Comment (ssoloff) [inline src/main/java/games/strategy/ui/SwingComponents.java]: When using the `JButton(Action)` constructor, as we do in `AbstractMovePanel`, the button text is initialized using the `Action` name.  Because we were not setting either the button text or the action name previously, the button ended up with an empty label.

`newAbstractAction()` was added as part of #2137, so I assumed there was a purpose for it. :smile:  Maybe it was added in a commit before the `swinglib` stuff started to take shape?  IDK, but its only caller is the one line in `AbstractMovePanel`, so, at worst, we can inline it there.

File: src/main/java/games/strategy/triplea/ui/AbstractMovePanel.java (hunk 2 above)
Comment (ssoloff) [inline src/main/java/games/strategy/triplea/ui/AbstractMovePanel.java]: There is no redundancy because we never explicitly set the cancel button text.  The label is only passed to `newAbstractAction()`, and, as described above, the `JButton` constructor uses the action name to initialize the button text.

I will investigate changing the `Action` field to a corresponding `JButton`.  I think that will work.

**Comment by ssoloff** [comment]
> Ideally, we could restructure the code to not have what is a seeming race condition as described with invokeLater().

I don't feel comfortable mucking with the `setActive()` method because I have no idea what other use cases it's intended to cover, and it has zero test coverage. :smile:

However, a possible fix for the race condition is to deterministically restore the cursor in `MovePanel#cleanUpSpecific()`, as is done in `MovePanel#cancelMoveAction()`.  `cleanUpSpecific()` appears to be the callback invoked upon move completion, regardless of whether it was cancelled or not.  I'll fiddle with that to see if it works.

**Comment by ssoloff** [comment]
@DanVanAtta The latest set of commits should address the points you raised above.  Please re-review when you get a chance.

File: src/main/java/games/strategy/triplea/ui/MovePanel.java [hunk 3]
```diff
@@ -1346,6 +1346,7 @@ protected final void cleanUpSpecific() {
     selectedUnits.clear();
     updateRouteAndMouseShadowUnits(null);
     forced = null;
+    getMap().showMouseCursor();
```
Comment (DanVanAtta) [inline src/main/java/games/strategy/triplea/ui/MovePanel.java]: I'm guessing this is a slightly unrelated fix, that at some point prior this call got dropped?

Comment (ssoloff) [inline src/main/java/games/strategy/triplea/ui/MovePanel.java]: Nope, this code is new (AFAICT going back to 2007 when `MapPanel#showMouseCursor()` was added).  This is my solution to the brittleness you mentioned above.  See my [previous comment](https://github.com/triplea-game/triplea/pull/2186#issuecomment-320782576) for an explanation.

Basically, now there may potentially be two calls to `getMap().showMouseCursor()` at the completion of a move from two different code paths.  But it should be less likely for a dev to get burned in the future by the roundabout nature of one of those paths that originate in `AbstractMovePanel#setActive()`.

I don't mind removing it if you think it just muddies the waters even more.  Its presence or absence will not affect the fix for #2185.

**Comment by ssoloff** [comment]
> I have not tested it though.

For anyone who might want to independently verify this fix, it's pretty easy to reproduce.

On `master`, play a game on any map.  After you combat move a unit to another territory, the cursor will still be hidden as long as it is over the map view.  Repeating the same steps on this branch, the cursor should be restored to its previously-visible state once you complete the move.

Also, for the other part of the fix, keep an eye on the Cancel button in the Move panel.  It should only be enabled while you are moving a unit.  Once you complete or cancel the move (via <kbd>ESC</kbd> or by clicking the Cancel button), the Cancel button should be disabled.

**Comment by ron-murhammer** [comment]
Works for me. Merging.
